harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -1553,20 +1553,30 @@ def train_batch(
         pass
 
     @abstractmethod
-    def build_training_data_loader(self) -> pytorch.DataLoader:
+    def build_training_data_loader(self) -> Union[pytorch.DataLoader, torch.utils.data.DataLoader]:
         """
         Defines the data loader to use during training.
 
-        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+        For full "determined" functionality, this must return an instance of
+        :py:class:`determined.pytorch.DataLoader`. It can also return an unwrapped
+        :py:class:`torch.utils.data.DataLoader` if you need more control over the underlying
+        DataLoader and are willing to sacrifice some determined features (ex: automatic data
+        sharding).
         """
         pass
 
     @abstractmethod
-    def build_validation_data_loader(self) -> pytorch.DataLoader:
+    def build_validation_data_loader(
+        self,
+    ) -> Union[pytorch.DataLoader, torch.utils.data.DataLoader]:
         """
         Defines the data loader to use during validation.
 
-        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+        For full "determined" functionality, this must return an instance of
+        :py:class:`determined.pytorch.DataLoader`. It can also return an unwrapped
+        :py:class:`torch.utils.data.DataLoader` if you need more control over the underlying
+        DataLoader and are willing to sacrifice some Determined features (ex: automatic data
+        sharding).
```

Comment:
Thank you for catching that our typing and docstring had fallen out of correctness.

However, this is a user-facing docstring, and given that I think this addition is much too vague.  What is `full "determined" functionality` anyway?

I wouldn't recommend answering that question here; I would link to [our existing docs on the subject](https://docs.determined.ai/latest/model-dev-guide/api-guides/apis-howto/api-pytorch-ug.html#customize-a-reproducible-dataset).

Something like:

```
Users with a MapDataset will normally return a :class:`determined.pytorch.DataLoader`, but users with an IterableDataset or with other advanced needs may return a bare ``torch.utils.data.DataLoader`` if they follow the steps described in :ref:`pytorch-reproducible-dataset`.
```

Reply:
I agree about the vagueness. It made me feel a little gross, too.

Ideally, the docstring should say _why_ a user would choose which class to return. It's somewhat easy to say "if you can't return a `det.pytorch.DataLoader`, return the other one. It's harder to come up with an explanation for "why managed" that's not vague.

I like your idea of referring to docs for that. That's something docs.determined.ai should do. The doc you suggested does have a nice note explaining how, I can't find anything on the site for why.

I think a voice chat is probably the best way to work something out from here. When you get to this and you've got the time to, could you please give me a call?

Reply:
Ended up with a little from column A, a little from column B.